includes/Traits/URL_Aware.php

```diff
@@ -0,0 +1,188 @@
+<?php
+/**
+ * Class WordPress\Plugin_Check\Traits\URL_Aware
+ *
+ * @package plugin-check
+ */
+
+namespace WordPress\Plugin_Check\Traits;
+
+trait URL_Aware {
+
+	/**
+	 * List of relevant global query variables to modify.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	private $query_globals = array(
+		'query_string',
+		'id',
+		'postdata',
+		'authordata',
+		'day',
+		'currentmonth',
+		'page',
+		'pages',
+		'multipage',
+		'more',
+		'numpages',
+		'pagenow',
+		'current_screen',
+	);
+
+	/**
+	 * Array of original values for the global state.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	private $global_values = array(
+		'get'         => array(),
+		'post'        => array(),
+		'server'      => array(),
+		'global_vars' => array(),
+	);
+
+	/**
+	 * Backups the original values for any global state that may be modified to be restored later.
+	 *
+	 * @since n.e.x.t
+	 */
+	protected function backup_globals() {
+		$this->global_values = array(
+			'get'    => $_GET,
+			'post'   => $_POST,
+			'server' => $_SERVER,
+		);
+
+		$global_vars = array();
+
+		foreach ( $this->query_globals as $query_global ) {
+			if ( isset( $GLOBALS[ $query_global ] ) ) {
+				$global_vars[ $query_global ] = $GLOBALS[ $query_global ];
+			}
+		}
+		if ( isset( $GLOBALS['wp_query'] ) ) {
+			$global_vars['wp_query'] = $GLOBALS['wp_query'];
+		}
+		if ( isset( $GLOBALS['wp_the_query'] ) ) {
+			$global_vars['wp_the_query'] = $GLOBALS['wp_the_query'];
+		}
+		if ( isset( $GLOBALS['wp'] ) ) {
+			$global_vars['wp'] = $GLOBALS['wp'];
+		}
```

Comment:
~~This logic is exactly the same as in the `foreach` loop above. Maybe we should just add those three keys to `$this->query_globals`?~~

**Update:** I see in `go_to()` they are handled differently. In that case, maybe we introduce another private property for this separate list of three globals? That way we can use the same `foreach` loop to go over both of these lists (by combining them before in a local variable).